### ISSUE TITLE
pyarrow.Array segfaults when constructed from generator without size param

### ISSUE THREAD

**Issue by cldellow** [apache/arrow#1687]
Hi! Thanks for making pyarrow & parquet-cpp -- this stuff is amazing!

I ran into an issue when getting started with pyarrow -- I was trying to use a generator to stream data into an array to avoid memory overhead from retaining the Python objects. I suspect this approach was wrong and I should instead have used `chunked_array`.

Still, I figured I'd ask: is it expected that the 3rd invocation segfaults?

```
import pyarrow as pa

def vals():
  for rownum in range(50):
    yield rownum

print('Making from array')
pa.array([x for x in vals()], type=pa.uint16())

print('Making from generator with size')
pa.array(vals(), type=pa.uint16(), size=50)

# This segfaults :(
print('Making from generator with no size')
pa.array(vals(), type=pa.uint16())
```

Environment: pyarrow 0.8.0, python 3.5.2.

**Comment by wesm** [comment]
This is definitely not the desired behavior! I think this may be fixed in master, but someone should confirm and open a JIRA if it is still broken (and make sure there is an appropriate unit test if there is not already)

**Comment by cldellow** [comment]
Thanks! I'll try to at least confirm if it's in master in the next couple weeks so this issue can be closed or replaced with a JIRA ticket.

**Comment by wesm** [comment]
I can confirm this is fixed on master, so will be in the 0.9.0 release
